source/common/upstream/load_balancer_impl.cc

```diff
@@ -92,42 +123,56 @@ void LoadBalancerBase::regenerateLocalityRoutingStructures() {
   // residual_capacity: 0 10000 15000
   // Now to find a locality to route (bucket) we could simply iterate over residual_capacity
   // searching where sampled value is placed.
-  residual_capacity_.resize(num_localities);
+  state.residual_capacity_.resize(num_localities);
 
   // Local locality (index 0) does not have residual capacity as we have routed all we could.
-  residual_capacity_[0] = 0;
+  state.residual_capacity_[0] = 0;
   for (size_t i = 1; i < num_localities; ++i) {
     // Only route to the localities that have additional capacity.
     if (upstream_percentage[i] > local_percentage[i]) {
-      residual_capacity_[i] =
-          residual_capacity_[i - 1] + upstream_percentage[i] - local_percentage[i];
+      state.residual_capacity_[i] =
+          state.residual_capacity_[i - 1] + upstream_percentage[i] - local_percentage[i];
     } else {
       // Locality with index "i" does not have residual capacity, but we keep accumulating previous
       // values to make search easier on the next step.
-      residual_capacity_[i] = residual_capacity_[i - 1];
+      state.residual_capacity_[i] = state.residual_capacity_[i - 1];
     }
   }
 };
 
-bool LoadBalancerBase::earlyExitNonLocalityRouting() {
-  if (host_set_.healthyHostsPerLocality().size() < 2) {
+void LoadBalancerBase::resizePerPriorityState(uint32_t size) {
+  size = std::max<uint32_t>(size, priority_set_.hostSetsPerPriority().size());
+  if (per_priority_state_.size() < size) {
+    for (size_t i = 0; i < size - per_priority_state_.size() + 1; ++i) {
```

Comment:
If the target `size` is 2 and `per_priority_state_.size()` is 1, then I end up growing `per_priority_state_.size()` by 2 here; or is it late at night and I'm not counting correctly?